Update documentation for adding new tests

Vineet1101 opened the pull request:
This PR updates the documentation for adding tests to the P4C repository.



---

### **Pull Request Title:**  
Update Documentation: Guide for Adding New Tests to P4C Repository

---

### **Description:**  
This PR adds a detailed guide to the documentation for developers looking to contribute tests to the P4C repository. The document outlines:  
1. Steps to add positive and negative tests.  
2. Instructions for handling expected output files.  
3. Guidelines for marking tests as "expected to fail" (Xfail).  
4. Considerations for different backends, tools, and test workflows.  
5. Recommendations for running tests via Bash scripts or CTest.  

The guide aims to enhance developer productivity and ensure consistency in test contributions.  

---

### **Motivation:**  
- Provide clarity on test organization and expected workflows.  
- Encourage best practices for adding, updating, and running tests.  
- Address gaps in existing documentation, particularly for backend-specific test requirements and Xfail behavior.  

---

### **Changes Made:**  
- Added a comprehensive document titled "Guide for Adding New Tests" under the `docs` directory.  
- Updated information on BMv2-specific file naming patterns and the behavior of Xfail tests in CI pipelines.  

---

### **Testing:**  
- Verified the document against existing repository structures and workflows.  
- Conducted a review to align with current testing practices and CI configurations.  

---

### **Checklist:**  
- [x] Documentation adheres to repository contribution guidelines.  
- [x] Added information is consistent with existing test workflows.  
- [x] Verified that all details are applicable to the current repository structure.  

---

### **Related Issues:**  
None directly. This PR aims to improve developer experience and test consistency proactively.  

---

Vineet1101 commented:
hey @jafingerhut can you please review this pr

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,143 @@
+
+3. **Maintaining Consistency**:
+
+   - Note that not all CI systems may treat Xfail tests the same way. For now, document the current behavior in the repository's CI documentation.
```
What do you mean not all "CI systems"?  Do you mean other CI systems in repositories other than p4c?  Or do you mean some p4c backends have CI tests that treat Xfail differently than other p4c backends do?

If it is that different p4c backends treat Xfail differently, then the sentence "For now, document the current behavior in the repository's CI documentation." is something that still remains to be done, yes?

If you do not know the answer yet, I think it would be best to replace that with something like:

"TODO: Document which p4c backend tests treat Xfail as described above, vs. those that treat Xfail differently."

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,143 @@
+     - `p4c/testdata/p4_16_samples`: General positive tests for the P4\_16 language.
+     - `p4c/testdata/p4_16_samples_outputs`: Expected outputs for these tests.
+     - Other directories may contain tests for specific backends or scenarios (e.g., BMv2 backend or PSA architecture).
+     - For BMv2-specific tests with packets, the source file name must match the pattern `*-bmv2.p4`.
```
Are there any other file name patterns that are important for test programs to follow?  It seems there are, from the output of the following command run in a clone of the p4c repo:
```
 find . | grep -i cmake | xargs grep 'testdata.*\*'
./backends/ebpf/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*_ebpf.p4"
./backends/ebpf/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*_ebpf-kernel.p4"
./backends/ebpf/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_ebpf_errors/*.p4"
./backends/tc/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4tc_samples/*.p4")
grep: ./backends/tofino/cmake: Is a directory
./backends/graphs/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_14_samples/*graph*.p4"
./backends/graphs/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*graph*.p4")
./backends/ubpf/CMakeLists.txt:set (UBPF_TEST_SUITES ${P4C_SOURCE_DIR}/testdata/p4_16_samples/*_ubpf.p4)
./backends/bmv2/CMakeLists.txt:set (P4TESTS_FOR_V1MODEL "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*-bmv2.p4")
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_bmv_errors/*-bmv2.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/dash/*-bmv2.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/fabric_*/fabric.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/omec/*.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/pins/*.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_14_samples/switch_*/switch.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_14_samples/*.p4"
./backends/bmv2/CMakeLists.txt:set (P4TESTS_FOR_PSA "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*-bmv2.p4")
[ ... more output lines omitted for bevity ... ]
```

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,143 @@
+
+     - `p4c/testdata/p4_16_samples`: General positive tests for the P4\_16 language.
+     - `p4c/testdata/p4_16_samples_outputs`: Expected outputs for these tests.
+     - Other directories may contain tests for specific backends or scenarios (e.g., BMv2 backend or PSA architecture).
```
It would be really nice if we could be exhaustive and describe what every single directory under `p4c/testdata` is for, precisely.

jafingerhut commented on docs/P4c-test-documentation.md:
````diff
@@ -0,0 +1,143 @@
+     ./build/p4test path/to/source_file.p4
+     ```
+   - This generates expected output files.
+   - Place the generated expected output files in the appropriate directory, such as `p4c/testdata/p4_16_samples_outputs`.
````
I don't think the command above will generate the expected output files needed to be copied into the p4_16_samples_outputs directory.

One way to do this is a command like `P4TEST_REPLACE=1 make check` that is mentioned in the top-level README, but that is to regenerate expected file for _all_ test cases, not only one.  I believe there is a way to run only one test and prefix it with `P4TEST_REPLACE=1 `, too, which would be good to mention here.

ChrisDodd commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,143 @@
+
+## **4. Marking Tests as "Expected to Fail" (Xfail)**
+
+An Xfail test is one that is expected to fail due to a known issue or limitation in the P4 compiler. The CI system handles these tests differently based on their type (positive or negative).
```
An XFail is a test that *should* pass (by the language spec and architecture), but currently fails due to known problems in the compiler.

jafingerhut commented on docs/P4c-test-documentation.md:
````diff
@@ -0,0 +1,143 @@
+
+   - If a change to the P4 compiler modifies the expected outputs, you can regenerate them by rerunning the tests with:
+     ```bash
+     ./build/p4test path/to/source_file.p4 --update-outputs
````
I can find no mention of `update-outputs` anywhere in the p4c repo.  Where did you find such an option?

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,143 @@
+
+### Differences Between Bash Scripts and CTest
+
+- **Options Available**: CTest may have additional or different options compared to Bash scripts. Check the documentation for details.
```
Regarding the statement "Check the documentation for details." it would be best to be more specific.  Check it _how_?  Meaning, is there a command someone can run to get the options available?  If so, show what they are here.  This _is_ the documentation that we want people to be reading to find out the details!

Vineet1101 commented:
@jafingerhut @ChrisDodd  I have made the changes as per your review. Please review them

jafingerhut commented on docs/P4c-test-documentation.md:
````diff
@@ -0,0 +1,205 @@
+     ```
+   - To generate or update the expected output file for a single test case, use:
+     ```bash
+     P4TEST_REPLACE=1 ./build/p4test path/to/source_file.p4
````
I can try out this command this weekend on a personal system of mine, but have you tried it to verify that it actually generates or updates the expected output files?  If so, I am surprised, because `p4test` has nothing built into it to tell which directory to put the expected output files in.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,205 @@
+     - **`p4_14_samples`**: Contains tests specific to the older P4\_14 language version.
+     - **`p4_14_samples_outputs`**: Stores the expected outputs for tests in `p4_14_samples`.
+     - **`bmv2`**: Holds tests targeting the BMv2 backend, including packet-processing logic and examples.
+     - **`psa`**: Includes tests for the PSA (Portable Switch Architecture) target.
```
PSA is not a target.  It is a P4 architecture.

In general, we could make a table with a row each target, and a column for each P4 architecture, and fill in the entries of the table with "yes" if that target has an implementation for that architecture, and "no" if it does not.  e.g. the P4 DPDK target implements the PSA and PNA architectures (at least partially), but it does not implement the v1model architecture at all.  The BMv2 back end implements the v1model architecture, and it has the beginnings of implementations of the PSA and PNA architectures, but they are missing a lot there.

I don't know off the top of my head which targets (if any) the `psa` directory has tests for.  I haven't looked at it recently, but it would be nice if we can find out and document it a little more precisely here.

jafingerhut replied:
Note: I don't mean by my previous comment that this PR _should_ have a table like I mentioned.  That was more background information for you, in case you were unaware of the relationship between P4 targets and architectxures.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,205 @@
+     - For eBPF-specific tests: `*_ebpf.p4` or `*_ebpf-kernel.p4`.
+     - For graph backend tests: `*graph*.p4`.
+     - For UBPF-specific tests: `*_ubpf.p4`.
+     - For PSA tests targeting BMv2: `*-bmv2.p4`.
```
There are many tests for the BMv2 back end using the v1model architecture that have file names matching `*-bmv2.p4`.  Do you have examples of specific files with names matching that pattern that use the PSA architecture, too?  There might be -- I have not checked.

jafingerhut replied:
Just did a quick check -- yes, there are many that use the PSA architecture, but even more that use the v1model architecture.

Vineet1101 commented:
hey @jafingerhut I can add todo their as the current reviews require me to do some little bit of research. So can you merge the pr I will add todo tag where changes are required as after it will be easy for me to make changes

fruffy commented:
> hey @jafingerhut I can add todo their as the current reviews require me to do some little bit of research. So can you merge the pr I will add todo tag where changes are required as after it will be easy for me to make changes

Our policy is not to merge PRs that are incorrect or flawed because there is no guarantee that the committer will follow up on their work. If @jafingerhut approves of the current version we are good to go but otherwise the changes need to be made in this PR.

jafingerhut commented:
> > hey @jafingerhut I can add todo their as the current reviews require me to do some little bit of research. So can you merge the pr I will add todo tag where changes are required as after it will be easy for me to make changes
> 
> Our policy is not to merge PRs that are incorrect or flawed because there is no guarantee that the committer will follow up on their work. If @jafingerhut approves of the current version we are good to go but otherwise the changes need to be made in this PR.

Sorry if this is turning out to be more work than originally expected -- perhaps my `smalltask` estimate in the project-ideas repo for this issue was a little bit low.

I am not quite at approval level yet on this one.  In particular, recommending commands that are claimed to do one thing, but do not, is just a recipe for readers trying out those commands, and getting confused and/or filing issues if the commands do not work as advertised.  Every command recommended in this article should be tested working as described, or the article should be adjusted until that is true.

Vineet1101 commented:
@jafingerhut  I have added the table as you have mentioned and also the name for bmv2-specific-test for both v1model as well as psa. I have removed that  " P4TEST_REPLACE=1 ./build/p4test path/to/source_file.p4"  for now as I will verify it by myself. Please review the pr and tell if any further changes are required in the documentation

fruffy commented:
> @fruffy @jafingerhut do I need to make any other changes in the docs please tell. I am happy to make any changes as per your review

The DCO check is still failing, you need to fix that. Please see these instructions: https://github.com/p4lang/p4c/blob/main/CONTRIBUTING.md#contributing-license

Vineet1101 commented:
@fruffy  i have resolved the issue of dco check

fruffy commented:
@jafingerhut Any further comments?

jafingerhut commented:
> @jafingerhut Any further comments?

Sorry, I know I've been doing a lot of p4lang reviews and PR recently, but not paying sufficient attention to this one yet.

I'd still like to try out every command suggested in this article, and verify that it does what the article says it does, before approving.

Vineet1101 commented:
@jafingerhut just wanted to know if there any changes that can be made

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+     - **`p4_16_samples_outputs`**: Stores the expected outputs for the tests in `p4_16_samples`.
+     - **`p4_14_samples`**: Contains tests specific to the older P4\_14 language version.
+     - **`p4_14_samples_outputs`**: Stores the expected outputs for tests in `p4_14_samples`.
+     - **`bmv2`**: Holds tests targeting the BMv2 backend, including packet-processing logic and examples.
```
Before this list of items, it implies this is a list of "multiple directories".  I know where to find the directories `p4_14_samples`, `p4_16_samples`, etc. listed above, but where are the directories `bmv2` and `psa` you are referring to?

jafingerhut replied:
Also `xdp`, `arch`, `error_samples`, `stf_samples`, `stress_tests`?

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+     - **`ebpf`**: Contains tests designed for the eBPF backend.
+     - **`xdp`**: Stores tests related to XDP programs using eBPF.
+     - **`arch`**: Contains architecture-specific tests, covering various P4 architectures.
+     - **`tools`**: Includes tests for tools provided in the P4C repository, such as `p4test` and `p4c-bm2-ss`.
```
I can find a `tools` directory, but it does not contain individual unit tests for `p4test` and `p4c-bm2-ss`.  It contains what I would call "test driver programs" (maybe there is a different name used somewhere in existing documentation), that are used by many, many test programs.  While it would be nice to mention them here, true, for the knowledge of p4c developers, most developers adding new test programs will not be modifying the files in the `tools` directory.

They _will_ commonly be adding files to subdirectories under the `p4c/testdata` directory, and I would hope that list of directories should all be mentioned in this list, and what they are for.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+2. **File Naming Conventions**:
+
+   - For **BMv2-specific tests** with packets, the source file name must match the pattern `*-bmv2.p4`.
+  - For **v1model architecture** tests: `*-bmv2-v1model.p4`.
```
There are currently _no_ test files with a file name that ends with `-bmv2-v1model.p4`.  If you tried to create one, it would not match the pattern `*-bmv2.p4`, and thus the cmakefiles and existing test infrastructure would _not_ run BMv2 on such a test program.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+   - For **BMv2-specific tests** with packets, the source file name must match the pattern `*-bmv2.p4`.
+  - For **v1model architecture** tests: `*-bmv2-v1model.p4`.
+
+  - For **PSA architecture** tests: `*-bmv2-psa.p4`.
```
There is exactly one existing test program `fwd-bmv2-psa.p4` that matches this file name pattern, and because it does NOT match the pattern `*-bmv2.p4`, I'm pretty sure that BMv2 is not run on this test program.  I do not know if that is a mistake on the part of the person who added that test program, or not.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+  - For **PSA architecture** tests: `*-bmv2-psa.p4`.
+  
+   - Additional file name patterns:
+     - For eBPF-specific tests: `*_ebpf.p4` or `*_ebpf-kernel.p4`.
```
OK, these file name suffixes I believe, because I find these two lines in the p4c repo:
```
./backends/ebpf/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*_ebpf.p4"
./backends/ebpf/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/*_ebpf-kernel.p4"
```
I think it is worth mentioning in this documentation which CMakeLists.txt files and lines like those are the _reason_ why these file name patterns are important.  Occasionally, someone might want to add a new suffix, e.g. for a new target, or a new P4 architecture.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+  
+   - Additional file name patterns:
+     - For eBPF-specific tests: `*_ebpf.p4` or `*_ebpf-kernel.p4`.
+     - For graph backend tests: `*graph*.p4`.
```
I cannot find any CMakeLists.txt file that has such a file name pattern.  If there is one, please mention it here.  If not, this is not a file name pattern that is significant to the test infrastructure, i.e. that affects what tests are run on a test program.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+   - Additional file name patterns:
+     - For eBPF-specific tests: `*_ebpf.p4` or `*_ebpf-kernel.p4`.
+     - For graph backend tests: `*graph*.p4`.
+     - For UBPF-specific tests: `*_ubpf.p4`.
```
I agree that this file name pattern should be here, because it is mentioned in this file, which I think should be added to this article as the reason why this file name pattern is significant:
```
./backends/ubpf/CMakeLists.txt:set (UBPF_TEST_SUITES ${P4C_SOURCE_DIR}/testdata/p4_16_samples/*_ubpf.p4)
```

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+     - For UBPF-specific tests: `*_ubpf.p4`.
+
+     - Other specialized tests under subdirectories such as:
+       - `p4tc_samples/` for TC tests.
```
The tests run on this directory are controlled by this occurrence in a CMakeLists.txt file, I think.  If correct, would be good to document that here:
```
./backends/tc/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4tc_samples/*.p4")
```

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+
+     - Other specialized tests under subdirectories such as:
+       - `p4tc_samples/` for TC tests.
+       - `fabric_*/` and `omec/` for Fabric and OMEC tests.
```
I am starting to get redundant with my comments, but in general, I think it is important the developers know the reason for why these directories are significant, and what causes the set of tests to be created, that are created.  Again, it is because of the contents of one or more CMakeLists.txt files.  For the `fabric_*' directories it is because of these lines, I think:
```
./backends/p4test/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/fabric_*/fabric.p4"
./backends/bmv2/CMakeLists.txt:  "${P4C_SOURCE_DIR}/testdata/p4_16_samples/fabric_*/fabric.p4"
```

jafingerhut commented on docs/P4c-test-documentation.md:
````diff
@@ -0,0 +1,235 @@
+
+   - If a change to the P4 compiler modifies the expected outputs, you can regenerate them by running:
+     ```bash
+     P4TEST_REPLACE=1 ./build/p4test path/to/source_file.p4
````
Have you personally tried this command?  If not, then why do you believe it does what you claim?

Part of writing documentation like this is to confirm yourself that what you recommend doing, actually does what you claim.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+
+1. **Source File Location**:
+
+   - Place the source file in the `p4c/testdata/p4_16_errors` directory.
```
That is not the only directory with negative tests.  Please list them all, and provide some guidance for developers to know which one to use for a particular test.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+1. **Adding the Xfail Mark**:
+
+   - Modify the test metadata to indicate that it is expected to fail.
+   - For example, you might add an `XFAIL` tag in the test description or a specific configuration file.
```
Please show at least one example, using an excerpt of a few current lines from one of the CMakeLists.txt files, that shows how a test is marked as XFAIL.  Please use an example that shows some comments for the tests, where those comments give a brief explanation of why the test is marked as XFAIL.

jafingerhut commented on docs/P4c-test-documentation.md:
````diff
@@ -0,0 +1,235 @@
+   - CTest provides an alternative way to execute tests, often integrated with the CI pipeline.
+   - Example command:
+     ```bash
+     ctest -R test_name
````
I have not tried using `P4TEST_REPLACE=1 ` prefixes on a `ctest` command, but that _might_ be one correct way to update the expected output files.  Trying it out and confirming it actually _does_ do so, would be a prerequisite for documenting it.

jafingerhut commented on docs/P4c-test-documentation.md:
```diff
@@ -0,0 +1,235 @@
+| DPDK     | ❌ | ✅ | ✅ |
+| eBPF     | ❌ | ✅ | ❌ |
+| UBPF     | ❌ | ✅ | ❌ |
+| P4Tools  | ✅ | ✅ | ❌ |
```
p4tools is more than one program, I thought?  e.g. p4testgen, p4smith, and maybe one or two others.  I do not know if it is true that all of them have the same support for P4 architectures as the others.

I tried to say in an earlier comment that I did not think it was necessary to include a table like this in this article.  I appreciate the attempt, but I suspect that "full support" claims of PSA for DPDK, eBPF, and uBPF might be overstated.  More complete than BMv2's PSA implementation, I easily believe, but for example I know that DPDK's PSA implementation normally does not do egress processing.  You can enable it, but I'm pretty sure it doesn't support PSA multicast, either.

Such details are probably best left for the detailed documentation of each specific p4c back end, not this article.  I'd recommend removing this table from this article.
